rapid_response_xblock/logger.py

```diff
@@ -27,58 +33,69 @@ class SubmissionRecorder(BaseBackend):
 
-    def send(self, event):
-        # TODO: feature flag whitelisting valid problems
-        if event['event_type'] == 'problem_check':
-            try:
-                user_id = event['context']['user_id']
-                problem_id = UsageKey.from_string(
+        Returns:
+             SubmissionEvent: The parsed submission event data (or None)
+        """
+        # Ignore if this event was not the submission of an answer
+        if event['event_type'] != 'problem_check':
+            return None
+        # Ignore if there were multiple submissions represented in this single event
+        event_submissions = event['event']['submission']
+        if len(event_submissions) > 1:
+            return None
+        submission_key, submission = event_submissions.items()[0]
+        # Ignore if the problem being answered has a blank submission or is not multiple choice
+        if not submission or submission.get('response_type') != MULTIPLE_CHOICE_TYPE:
+            return None
+
+        try:
+            return SubmissionEvent(
+                raw_data=event,
+                user_id=event['context']['user_id'],
+                problem_usage_key=UsageKey.from_string(
                     event['event']['problem_id']
-                )
-                # problem_id.course_id may have a missing run
-                # so we need to grab the course key separately
-                course_key = CourseLocator.from_string(
+                ),
+                course_key=CourseLocator.from_string(
                     event['context']['course_id']
-                )
-
-                # This is assuming the event only includes information
-                # for one answer at a time
-                submission = event['event']['submission']
-                keys = list(submission.keys())
-                if len(keys) != 1:
-                    # Not sure if this case is possible
-                    raise Exception(
-                        "Expected only one answer in problem_check event"
-                    )
+                ),
+                answer_text=submission['answer'],
+                answer_id=event['event']['answers'][submission_key]
+            )
+        except:  # pylint: disable=bare-except
+            log.exception("Unable to parse event data as a submission: %s", event)
 
-                submission_key = keys[0]
-
-                answer_text = submission[submission_key]['answer']
-                answer_id = event['event']['answers'][submission_key]
-
-                open_run = RapidResponseRun.objects.filter(
-                    problem_usage_key=problem_id,
-                    course_key=course_key,
-                    open=True
-                ).first()
-                if not open_run:
-                    # Problem is not open
-                    return
+    def send(self, event):
+        sub = self.parse_submission_event(event)
```

Comment:
Is any functionality changing here or are you just breaking it into two functions for code cleanliness?

Reply:
Mostly cleanliness/clarity, but I also limited the scope of the `try` block. This only catches and logs exceptions when we're trying to parse an event that we think is a multiple choice problem response. Any exceptions that occur while we query/create objects for our models are not part of the same `try` block. It also doesn't raise when there is more than 1 problem type. This just ignores that case.